pkg/scheduler/objects/preemption.go

```diff
@@ -658,6 +660,10 @@ func (pcr *predicateCheckResult) populateVictims(victimsByNode map[string][]*All
 	for i := 0; i <= pcr.index; i++ {
 		victim := victimList[i]
 		pcr.victims = append(pcr.victims, victim)
+		// Take the instance type info from the first victim Allocation of this given nodeID
```

Comment:
This is just incorrect. Just because preemptions happened on a node does not ensure that node will ultimately be where the new allocation goes.

Reply:
The instance type of the node used for the NEW allocation is what's needed. This has nothing at all to do with the preemption logic. Please fix this.

Reply:
Sorry I did not make it clear previously, 

the `tryNodes()` function below is called by 

```
func (p *Preemptor) TryPreemption() (*Allocation, bool) {
...
        // try to find a node to schedule on and victims to preempt
	nodeID, instType, victims, ok := p.tryNodes() // The nodeId, instType info returned here is used to construct
                                                                                     //  a new allocation below
	if !ok {
		// no preemption possible
		return nil, false
	}
...
	if p.headRoom.FitInMaxUndef(p.ask.GetAllocatedResource()) {
		log.Logger().Info("Reserving node for ask after preemption",
			zap.String("allocationKey", p.ask.GetAllocationKey()),
			zap.String("nodeID", nodeID),
			zap.Int("victimCount", len(victims)))
		return newReservedAllocation(Reserved, nodeID, instType, p.ask), true
	}

...

```

which in turn is called by

```
// tryAllocate will perform a regular allocation of a pending request, includes placeholders.
func (sa *Application) tryAllocate(headRoom *resources.Resource, preemptionDelay time.Duration, preemptAttemptsRemaining *int, nodeIterator func() NodeIterator, fullNodeIterator func() NodeIterator, getNodeFn func(string) *Node) *Allocation {

```
which returns Allocation object, which need the instance type info.

```
func (p *Preemptor) tryNodes() (string, string, []*Allocation, bool) {
	// calculate victim list for each node
	predicateChecks := make([]*si.PreemptionPredicatesArgs, 0)
	victimsByNode := make(map[string][]*Allocation)
	for nodeID, nodeAvailable := range p.nodeAvailableMap {
		allocations, ok := p.allocationsByNode[nodeID]
		if !ok {
			// no allocations present, but node may still be available for scheduling
			allocations = make([]*Allocation, 0)
		}
		// identify which victims and in which order should be tried
		if idx, victims := p.calculateVictimsByNode(nodeAvailable, allocations); victims != nil {
			victimsByNode[nodeID] = victims
			keys := make([]string, 0)
			for _, victim := range victims {
				keys = append(keys, victim.GetAllocationKey())
			}
			// only check this node if there are victims or we have not already tried scheduling
			if len(victims) > 0 || !p.nodesTried {
				predicateChecks = append(predicateChecks, &si.PreemptionPredicatesArgs{
					AllocationKey:         p.ask.GetAllocationKey(),
					NodeID:                nodeID,
					PreemptAllocationKeys: keys,
					StartIndex:            int32(idx),
				})
			}
		}
	}
	// call predicates to evaluate each node
	result := p.checkPreemptionPredicates(predicateChecks, victimsByNode)
	if result != nil && result.success {
		return result.nodeID, result.instType, result.victims, true
	}

	return "", UnknownInstanceType, nil, false
}
```

which calls

```
	// call predicates to evaluate each node
	result := p.checkPreemptionPredicates(predicateChecks, victimsByNode)
	if result != nil && result.success {
		return result.nodeID, result.instType, result.victims, true
	}
```

My understanding is that we choose an candidate node (after identifying preemption vicitims) for new allocation, so I need to make 

`         result := p.checkPreemptionPredicates(predicateChecks, victimsByNode)
`
populate the instanceType of the successfully chosen node candidate. 

Wonder if we can have a screen share session for us to walk through a bit. 

Thanks a lot.

Reply:
The Preemptor stores mappings between node and various information needed for preemption. See initWorkingState() -- you can add a nodeInstanceTypeMap there and populate it. Then when a node is chosen for the allocation, look it up in that map. That's far simpler than what you have done here.

Reply:
It was via a discussion with @wilfred-s @yangwwei that we decided to add an instanceType in Allocation object, because we don't have a global map between nodeId and instanceType.  We not only need instanceType for preemption code (which is new), but also need instanceType for all nodes.  I wish we have a global map, but we think it's more of a management hassle, so we decided to have the instType info included together with nodeId when creating Allocation object.

Reply:
The down side is with my current implementation is that whenever we create an Allocation object, we need to have both the nodeId and instType information available.

Reply:
We do not need the instance type for all nodes, only those that might be candidates. The code in initWorkingState() that can be updated:

```
allocationsByNode := make(map[string][]*Allocation)
queueByAlloc := make(map[string]*QueuePreemptionSnapshot)
nodeAvailableMap := make(map[string]*resources.Resource)
nodeInstanceTypeMap := make(map[string]string) // <-- TODO add this

// build a map from NodeID to allocation and from allocationID to queue capacities
for _, victims := range p.allocationsByQueue {
	for _, allocation := range victims.PotentialVictims {
		nodeID := allocation.GetNodeID()
		allocations, ok := allocationsByNode[nodeID]
		if !ok {
			allocations = make([]*Allocation, 0)
		}
		allocationsByNode[nodeID] = append(allocations, allocation)
		queueByAlloc[allocation.GetAllocationKey()] = victims
		nodeInstanceTypeMap[nodeID] = node.GetInstanceType() // <-- TODO add this
	}
}
```

Reply:
Thanks, when I said "all nodes", I meant outside the preemption code, when we do an allocation, we need similar information. Doing a partial map for preemption code makes it a bit different than the handling outside the preemption code. If you think this is what we should do, I will give it a try.

Do you think it's worth to have a global nodeId to instanceType mapping? we discussed this before and chose to use the alternative solution that's currently in the PR.

Thanks.

Reply:
I think the solution I suggested should be sufficient for now. A global mapping further complicates the context, and requires additional locking at multiple levels, which the node iterator prevents currently. This is something that could be revisited in the future, but that's where we are at today.

Reply:
Thanks a lot @craigcondit just updated with new rev. In the code change you suggested, I had to do:

nodeInstanceTypeMap[nodeID] = allocation.GetInstanceType()

instead of
 
nodeInstanceTypeMap[nodeID] = node.GetInstanceType() 

Would you please take a look again when you get chance? thanks

Reply:
Added review comments - I think this can work, but we need to add some defensive checks in case the instance type is somehow not populated.

Reply:
Many thanks @craigcondit good point, I just updated the PR with a few changes, including your above comment,  assigning the map

`p.nodeInstanceTypeMap = nodeInstanceTypeMap`

which I forgot in last rev, and using "UNKNOWN", 

Thanks again for taking further look!